### PR TITLE
fix: Fix SortBuffer ensureOutputFits estimateOutputSize inaccurate

### PR DESCRIPTION
**Pull request by jinchengchenghh**
The output batch reserved size should be rowSize * numRows, missed numRows before.

### CODE REVIEW

**Comment by FelixYBW** [comment]
Issue is not fixed:

```
Error Source: RUNTIME
Error Code: INVALID_STATE
Reason: Operator::getOutput failed for [operator: OrderBy, plan node ID: 1]: Error during calling Java code from native code: org.apache.gluten.memory.memtarget.ThrowOnOomMemoryTarget$OutOfMemoryException: Not enough spark off-heap execution memory. Acquired: 8.0 MiB, granted: 4.0 MiB. Try tweaking config option spark.memory.offHeap.size to get larger space to run this application (if spark.gluten.memory.dynamic.offHeap.sizing.enabled is not enabled). 
Current config settings: 
	spark.gluten.memory.offHeap.size.in.bytes=8.3 GiB
	spark.gluten.memory.task.offHeap.size.in.bytes=8.3 GiB
	spark.gluten.memory.conservative.task.offHeap.size.in.bytes=4.2 GiB
	spark.memory.offHeap.enabled=true
	spark.gluten.memory.dynamic.offHeap.sizing.enabled=false
Memory consumer stats: 
	Task.139909:                                              Current used bytes:  8.3 GiB, peak bytes:        N/A
	\- Gluten.Tree.61:                                        Current used bytes:  8.3 GiB, peak bytes:    8.3 GiB
	   \- root.61:                                            Current used bytes:  8.3 GiB, peak bytes:    8.3 GiB
	      +- NativePlanEvaluator-61.0:                        Current used bytes:  8.3 GiB, peak bytes:    8.3 GiB
	      |  \- single:                                       Current used bytes:  8.3 GiB, peak bytes:    8.3 GiB
	      |     +- root:                                      Current used bytes:  8.3 GiB, peak bytes:    8.3 GiB
	      |     |  +- task.Gluten_Stage_2_TID_139909_VTID_61: Current used bytes:  8.3 GiB, peak bytes:    8.3 GiB
	      |     |  |  +- node.1:                              Current used bytes:  8.3 GiB, peak bytes:    8.3 GiB
	      |     |  |  |  \- op.1.0.0.OrderBy:                 Current used bytes:  8.3 GiB, peak bytes:    8.3 GiB
	      |     |  |  +- node.2:                              Current used bytes: 96.0 KiB, peak bytes: 1024.0 KiB
	      |     |  |  |  \- op.2.0.0.Window:                  Current used bytes: 96.0 KiB, peak bytes:   96.0 KiB
	      |     |  |  +- node.3:                              Current used bytes:    0.0 B, peak bytes:      0.0 B
	      |     |  |  |  \- op.3.0.0.FilterProject:           Current used bytes:    0.0 B, peak bytes:      0.0 B
	      |     |  |  \- node.0:                              Current used bytes:    0.0 B, peak bytes:      0.0 B
	      |     |  |     \- op.0.0.0.ValueStream:             Current used bytes:    0.0 B, peak bytes:      0.0 B
	      |     |  \- default_leaf:                           Current used bytes:    0.0 B, peak bytes:      0.0 B
	      |     \- gluten::MemoryAllocator:                   Current used bytes:    0.0 B, peak bytes:      0.0 B
	      +- ArrowContextInstance.6:                          Current used bytes:  8.0 MiB, peak bytes:    8.0 MiB
	      +- IteratorMetrics.61:                              Current used bytes:    0.0 B, peak bytes:      0.0 B
	      |  \- single:                                       Current used bytes:    0.0 B, peak bytes:      0.0 B
	      |     +- root:                                      Current used bytes:    0.0 B, peak bytes:      0.0 B
	      |     |  \- default_leaf:                           Current used bytes:    0.0 B, peak bytes:      0.0 B
	      |     \- gluten::MemoryAllocator:                   Current used bytes:    0.0 B, peak bytes:      0.0 B
	      +- IndicatorVectorBase#init.61.OverAcquire.0:       Current used bytes:    0.0 B, peak bytes:      0.0 B
	      +- ShuffleReader.3.OverAcquire.0:                   Current used bytes:    0.0 B, peak bytes:      0.0 B
	      +- IndicatorVectorBase#init.61:                     Current used bytes:    0.0 B, peak bytes:      0.0 B
	      |  \- single:                                       Current used bytes:    0.0 B, peak bytes:      0.0 B
	      |     +- root:                                      Current used bytes:    0.0 B, peak bytes:      0.0 B
	      |     |  \- default_leaf:                           Current used bytes:    0.0 B, peak bytes:      0.0 B
	      |     \- gluten::MemoryAllocator:                   Current used bytes:    0.0 B, peak bytes:      0.0 B
	      +- IteratorMetrics.61.OverAcquire.0:                Current used bytes:    0.0 B, peak bytes:      0.0 B
	      +- NativePlanEvaluator-61.0.OverAcquire.0:          Current used bytes:    0.0 B, peak bytes:      0.0 B
	      \- ShuffleReader.3:                                 Current used bytes:    0.0 B, peak bytes:   16.0 MiB
	         \- single:                                       Current used bytes:    0.0 B, peak bytes:   16.0 MiB
	            +- root:                                      Current used bytes:    0.0 B, peak bytes: 1024.0 KiB
	            |  \- default_leaf:                           Current used bytes:    0.0 B, peak bytes:  579.8 KiB
	            \- gluten::MemoryAllocator:                   Current used bytes:    0.0 B, peak bytes:  401.3 KiB

	at org.apache.gluten.memory.memtarget.ThrowOnOomMemoryTarget.borrow(ThrowOnOomMemoryTarget.java:105)
	at org.apache.gluten.memory.listener.ManagedReservationListener.reserve(ManagedReservationListener.java:49)
	at org.apache.gluten.vectorized.ColumnarBatchOutIterator.nativeHasNext(Native Method)
	at org.apache.gluten.vectorized.ColumnarBatchOutIterator.hasNext0(ColumnarBatchOutIterator.java:57)
	at org.apache.gluten.iterator.ClosableIterator.hasNext(ClosableIterator.java:39)
	at scala.collection.convert.Wrappers$JIteratorWrapper.hasNext(Wrappers.scala:43)
	at org.apache.gluten.iterator.IteratorsV1$InvocationFlowProtection.hasNext(IteratorsV1.scala:159)
	at org.apache.gluten.iterator.IteratorsV1$IteratorCompleter.hasNext(IteratorsV1.scala:71)
	at org.apache.gluten.iterator.IteratorsV1$PayloadCloser.hasNext(IteratorsV1.scala:37)
	at org.apache.gluten.iterator.IteratorsV1$LifeTimeAccumulator.hasNext(IteratorsV1.scala:100)
	at scala.collection.Iterator.isEmpty(Iterator.scala:385)
	at scala.collection.Iterator.isEmpty$(Iterator.scala:385)
	at org.apache.gluten.iterator.IteratorsV1$LifeTimeAccumulator.isEmpty(IteratorsV1.scala:90)
	at org.apache.gluten.execution.VeloxColumnarToRowExec$.toRowIterator(VeloxColumnarToRowExec.scala:121)
	at org.apache.gluten.execution.VeloxColumnarToRowExec.$anonfun$doExecuteInternal$1(VeloxColumnarToRowExec.scala:77)
	at org.apache.spark.rdd.RDD.$anonfun$mapPartitions$2(RDD.scala:949)
	at org.apache.spark.rdd.RDD.$anonfun$mapPartitions$2$adapted(RDD.scala:949)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:374)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:338)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:374)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:338)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:374)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:338)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:374)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:338)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.run(Task.scala:131)
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:506)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1471)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:509)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:750)

```

**Comment by FelixYBW** [comment]
@JkSelf @zhztheplayer 

Does the fix make sense? Where the 1.2 comes from?

`  const uint64_t outputBufferSizeToReserve =
      estimatedOutputRowSize_.value() * 1.2;
`

**Comment by JkSelf** [comment]
@FelixYBW @jinchengchenghh @zhztheplayer 
1.2 is experience value from https://github.com/facebookincubator/velox/blob/main/velox/exec/GroupingSet.cpp#L926.  Can you help to check whether the code is return [here ](https://github.com/facebookincubator/velox/blob/main/velox/exec/SortBuffer.cpp#L263)and not trigger spill?

File: velox/exec/SortBuffer.cpp [hunk 1]
```diff
@@ -263,8 +266,14 @@ void SortBuffer::ensureOutputFits() {
...
 
+  const vector_size_t numRows =
+      output_ == nullptr ? batchSize : std::max(0, batchSize - output_->size());
+  if (numRows == 0) {
+    return;
+  }
+
   const uint64_t outputBufferSizeToReserve =
-      estimatedOutputRowSize_.value() * 1.2;
+      estimatedOutputRowSize_.value() * numRows;
```
Comment (zhztheplayer) [inline velox/exec/SortBuffer.cpp:276]: If previously we had 1.2 as factor, why not to use `estimatedOutputRowSize_.value() * numRow * 1.2`?

**Comment by zhztheplayer** [comment]
> @FelixYBW @jinchengchenghh @zhztheplayer 1.2 is experience value from https://github.com/facebookincubator/velox/blob/main/velox/exec/GroupingSet.cpp#L926. Can you help to check whether the code is return [here ](https://github.com/facebookincubator/velox/blob/main/velox/exec/SortBuffer.cpp#L263)and not trigger spill?

BTW I think we may need to add some information in OOM message to indicate shrink / spill history to help trouble shooting.

File: velox/exec/SortBuffer.cpp (hunk 1 above)
Comment (jinchengchenghh) [reply velox/exec/SortBuffer.cpp:276]: I don't think we need the factor, the estimate size is for row format, but the output is column format, the size of latter is smaller than former. And I test the `TEST_F(RowContainerTest, estimateRowSize)`, compare the data size with rowVector->estimateFlatSize(), row format is 64 while column format is 45 when the number of rows is 200,000 and data type is (BIGINT, VARCHAR).

**Comment by jinchengchenghh** [comment]
Yes, I agree, this spill flag should return to OperatorStats. Now I will only print the spill information to log.

File: velox/exec/SortBuffer.cpp (hunk 1 above)
Comment (zhztheplayer) [reply velox/exec/SortBuffer.cpp:276]: Thank for the explanation. Though I think for this kind of estimation we may take the worst case into account? Say if the batch only has 100 rows or even less. Is the calculation still safe in the case?

All based on the assumption that when the calculation fails, OOM will definitely take place which is unexpected.

Comment (jinchengchenghh) [reply velox/exec/SortBuffer.cpp]: Yes, addressed.

File: velox/exec/SortBuffer.cpp [hunk 2]
```diff
@@ -259,15 +262,24 @@ void SortBuffer::ensureOutputFits() {
...
     return;
   }
 
   const uint64_t outputBufferSizeToReserve =
-      estimatedOutputRowSize_.value() * 1.2;
+      estimatedOutputRowSize_.value() * numRows * 1.2;
   {
     memory::ReclaimableSectionGuard guard(nonReclaimableSection_);
     if (pool_->maybeReserve(outputBufferSizeToReserve)) {
+      if (spiller_ != nullptr) {
```
Comment (xiaoxmeng) [inline velox/exec/SortBuffer.cpp:280]: Let's remove this log? Does this help debug?

File: velox/exec/SortBuffer.cpp [hunk 3]
```diff
@@ -259,15 +262,24 @@ void SortBuffer::ensureOutputFits() {
     return;
   }
 
-  if (!estimatedOutputRowSize_.has_value()) {
+  if (!estimatedOutputRowSize_.has_value() || spiller_ != nullptr) {
+    return;
+  }
+
+  const vector_size_t numRows =
```
Comment (xiaoxmeng) [inline velox/exec/SortBuffer.cpp:269]: Can we simply check on batchSize? The output buffer might not be able to reuse?

**Review by xiaoxmeng — COMMENTED** [review]
@jinchengchenghh LGTM. Thanks for the fix % minors.

File: velox/exec/SortBuffer.cpp (hunk 2 above)
Comment (xiaoxmeng) [reply velox/exec/SortBuffer.cpp:275]: Then let's remove this?

File: velox/exec/SortBuffer.cpp [hunk 4]
```diff
@@ -259,15 +262,19 @@ void SortBuffer::ensureOutputFits() {
     return;
   }
 
-  if (!estimatedOutputRowSize_.has_value()) {
+  if (!estimatedOutputRowSize_.has_value() || spiller_ != nullptr ||
+      batchSize == 0) {
```
Comment (xiaoxmeng) [inline velox/exec/SortBuffer.cpp:266]: batch size can't be zero and can you just add VELOX_CHECK at the entrance of this function? Thanks!

File: velox/exec/SortBuffer.cpp (hunk 1 above)
Comment (FelixYBW) [reply velox/exec/SortBuffer.cpp]: If we need the 1.2 factor, why don't we put it in higher level like estimatedOutputRowSize_'s calculation to make sure all estimation is the over estimated? Also we need to document it as a guideline.

**Comment by FelixYBW** [comment]
> nk we may need to add some information in OOM message to indicate shrink / spill history to help troubl

Can you add it? also is it possible to take the callstack trace to the log?

File: velox/exec/SortBuffer.cpp (hunk 4 above)
Comment (jinchengchenghh) [reply velox/exec/SortBuffer.cpp:266]: `const vector_size_t batchSize =
      std::min<uint64_t>(numInputRows_ - numOutputRows_, maxOutputRows)`, The batchSize maybe 0 if we input an empty RowVector.

**Comment by jinchengchenghh** [comment]
> > nk we may need to add some information in OOM message to indicate shrink / spill history to help troubl
> 
> Can you add it? also is it possible to take the callstack trace to the log?

I don't know how to get the shrink/spill history when OOM. We could get the callstack trace by debug version, but we cannot get much information in release version. CC @JkSelf She gets the stack before.

**Comment by jinchengchenghh** [comment]
Can we get the spill information in Spark UI, the Operator statistic should contain some information about spill

**Review by xiaoxmeng — APPROVED** [review]
@jinchengchenghh LGTM. Thanks for the update!

**Review by xiaoxmeng — COMMENTED** [review]
@jinchengchenghh CI is red. Thanks!

**Comment by jinchengchenghh** [comment]
CI issue https://github.com/facebookincubator/velox/issues/11550 and https://github.com/facebookincubator/velox/issues/11567

**Comment by jinchengchenghh** [comment]
Discussed with @zhztheplayer offline, we catch std::exception not VeloxException in JNI_METHOD_END, otherwise, we can call VeloxException.stacktrace to get the stack. CC @FelixYBW 
This is velox exception printed message in velox release version
```


File: /home/runner/work/velox/velox/velox/velox/exec/tests/TaskTest.cpp
Line: 2200
Stack trace:
# 0  _ZN8facebook5velox7process10StackTraceC1Ei
# 1  _ZN8facebook5velox14VeloxException5State4makeIZNS1_C4EPKcmS5_St17basic_string_viewIcSt11char_traitsIcEES9_S9_S9_bNS1_4TypeES9_EUlRT_E_EESt10shared_ptrIKS2_ESA_SB_
# 2  _ZN8facebook5velox14VeloxExceptionC1EPKcmS3_St17basic_string_viewIcSt11char_traitsIcEES7_S7_S7_bNS1_4TypeES7_
# 3  _ZN8facebook5velox17VeloxRuntimeErrorC2EPKcmS3_St17basic_string_viewIcSt11char_traitsIcEES7_S7_S7_bS7_
# 4  _ZN8facebook5velox6detail14veloxCheckFailINS0_17VeloxRuntimeErrorERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEEvRKNS1_18VeloxCheckFailArgsET0_
# 5  _ZZN8facebook5velox4exec4test32TaskTest_taskReclaimFailure_Test8TestBodyEvENKUlPNS1_7SpillerEE_clES5_
# 6  _ZSt13__invoke_implIvRZN8facebook5velox4exec4test32TaskTest_taskReclaimFailure_Test8TestBodyEvEUlPNS2_7SpillerEE_JS6_EET_St14__invoke_otherOT0_DpOT1_
# 7  
```

**Comment by jinchengchenghh** [comment]
We catch the exception by `e.what`, it is https://github.com/facebookincubator/velox/blob/main/velox/common/base/VeloxException.cpp#L261, and it should print the stacktrace message if have, Function `isStackTraceEnabled()` https://github.com/facebookincubator/velox/blob/main/velox/common/base/VeloxException.cpp#L134 will check if the exception has stack. And does your version contains this PR  https://github.com/apache/incubator-gluten/pull/2398 to enable print user exception?  @FelixYBW

**Comment by FelixYBW** [comment]
> Can we get the spill information in Spark UI, the Operator statistic should contain some information about spill

I can see spill info from my UI

**Comment by FelixYBW** [comment]
> lox

It's not the correct call stack trace. It's too late when we saw the OOM issue. I think we have to pass the call stack info along with each memory request. Since Spark doesn't allow such additional info, we have to handle from native.

**Pull request closed without merging**